File: go/models/defense.go
package models

import "time"

type PaladinStats struct {
	ActiveLayers int              `json:"active_layers"`
	TotalChecks  int              `json:"total_checks"`
	BlockRate    float64          `json:"block_rate"`
	Layers       []LayerStatEntry `json:"layers"`
}

type LayerStatEntry struct {
	Name    string  `json:"name"`
	Enabled bool    `json:"enabled"`
	Checks  int     `json:"checks"`
	Blocked int     `json:"blocked"`
	AvgMs   float64 `json:"avg_ms"`
}

type EnableLayerRequest struct {
	Layer   string `json:"layer"`
	Enabled bool   `json:"enabled"`
}

type EnableLayerResponse struct {
	Layer   string `json:"layer"`
	Enabled bool   `json:"enabled"`
}

type TrustValidateRequest struct {
	Input     string            `json:"input"`
	Source    string            `json:"source,omitempty"`
	Context   map[string]string `json:"context,omitempty"`
}

type TrustValidateResponse struct {
	Trusted    bool    `json:"trusted"`
	Score      float64 `json:"score"`
	Reason     string  `json:"reason,omitempty"`
	Source     string  `json:"source"`
}

type TrustProfile struct {
	Source       string    `json:"source"`
	TrustScore   float64  `json:"trust_score"`
	Interactions int       `json:"interactions"`
	LastSeen     time.Time `json:"last_seen"`
	Flags        []string  `json:"flags,omitempty"`
}

type RagDetectRequest struct {
	Query     string   `json:"query"`
	Documents []string `json:"documents"`
	Response  string   `json:"response,omitempty"`
}

type RagDetectResponse struct {
	Poisoned   bool               `json:"poisoned"`
	Confidence float64            `json:"confidence"`
	Flagged    []int              `json:"flagged"`
	Details    map[string]float64 `json:"details"`
}

type CircuitBreakerEvalRequest struct {
	Service   string  `json:"service"`
	ErrorRate float64 `json:"error_rate,omitempty"`
}

type CircuitBreakerResponse struct {
	Service  string `json:"service"`
	State    string `json:"state"`
	Failures int    `json:"failures"`
	LastFail string `json:"last_failure,omitempty"`
}

type AdaptiveEvalRequest struct {
	Input    string            `json:"input"`
	Context  map[string]string `json:"context,omitempty"`
}

type AdaptiveEvalResponse struct {
	Decision string  `json:"decision"`
	Score    float64 `json:"score"`
	Adapted  bool    `json:"adapted"`
	Model    string  `json:"model"`
}

type AdaptiveLearnRequest struct {
	Feedback string `json:"feedback"`
	Label    string `json:"label"`
	InputID  string `json:"input_id"`
}

type AdaptiveLearnResponse struct {
	Accepted bool   `json:"accepted"`
	Message  string `json:"message"`
}

File: rust/src/models/defense.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct TrustValidateRequest {
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trust_level: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct TrustValidateResponse {
    #[serde(default)]
    pub trusted: bool,
    #[serde(default)]
    pub trust_score: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub violations: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct TrustProfile {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_id: Option<String>,
    #[serde(default)]
    pub trust_score: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trust_level: Option<String>,
    #[serde(default)]
    pub total_requests: i64,
    #[serde(default)]
    pub blocked_requests: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct RagDetectRequest {
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct RagDetectResponse {
    #[serde(default)]
    pub poisoning_detected: bool,
    #[serde(default)]
    pub confidence: f64,
    #[serde(default)]
    pub poisoned_sources: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct RagSecureQueryRequest {
    pub query: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sources: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct RagSecureQueryResponse {
    #[serde(default)]
    pub safe: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sanitized_query: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filtered_sources: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct CircuitBreakerRequest {
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct CircuitBreakerResponse {
    #[serde(default)]
    pub state: String,
    #[serde(default)]
    pub allowed: bool,
    #[serde(default)]
    pub failure_count: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub threshold: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reset_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct AdaptiveEvalRequest {
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct AdaptiveEvalResponse {
    #[serde(default)]
    pub decision: String,
    #[serde(default)]
    pub confidence: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub adapted_threshold: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub learning_rate: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct PaladinStats {
    #[serde(default)]
    pub total_evaluations: i64,
    #[serde(default)]
    pub blocked: i64,
    #[serde(default)]
    pub approved: i64,
    #[serde(default)]
    pub layers: Vec<PaladinLayerStats>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct PaladinLayerStats {
    pub name: String,
    #[serde(default)]
    pub enabled: bool,
    #[serde(default)]
    pub total_checks: i64,
    #[serde(default)]
    pub blocks: i64,
    #[serde(default)]
    pub avg_latency_ms: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct GuardNetRequest {
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub threshold: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct GuardNetResponse {
    #[serde(default)]
    pub jailbreak_probability: f64,
    #[serde(default)]
    pub should_block: bool,
    #[serde(default)]
    pub hierarchy_scores: Vec<serde_json::Value>,
    #[serde(default)]
    pub latency_ms: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct UnifiedDefenseRequest {
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub threshold: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct UnifiedDefenseResponse {
    #[serde(default)]
    pub should_block: bool,
    #[serde(default)]
    pub overall_score: f64,
    #[serde(default)]
    pub results: serde_json::Value,
    #[serde(default)]
    pub latency_ms: f64,
}
